src/backend/cdb/cdbmutate.c

```diff
@@ -1113,6 +1116,375 @@ request_explicit_motion(Plan *plan, Index resultRelationsIdx, List *rtable)
 	plan->flow->segidColIdx = segidColIdx;
 }
 
+typedef struct
+{
+	plan_tree_base_prefix	base;	/* Required prefix for
+									 * plan_tree_walker/mutator */
+
+	List					*absent_vars;
+	Index					result_relation_idx;
+	bool					found;
+} append_absent_targetlist_context;
+
+static void
+append_absent_targetlist_mutator(Plan *plan,
+								 append_absent_targetlist_context *pcontext)
+{
+	Node	*node;
+	int		targetResno;
+
+	node = (Node *) plan;
+
+	Assert(is_plan_node(node));
+
+	/*
+	 * GPDB_90_MERGE_FIXME: We also need to change width and cost here. But since the plan has been
+	 * generated at this stage, it is not clear how we could recalculate the cost.
+	 */
+	if (node->type >= T_SeqScan && node->type <=T_WorkTableScan)
+	{
+		Scan		*scan;
+		ListCell	*lcv;
+
+		scan = (Scan *) node;
+		if (scan->scanrelid != pcontext->result_relation_idx)
+			return;
+
+		targetResno = list_length(plan->targetlist) + 1;
+		foreach(lcv, pcontext->absent_vars)
+		{
+			TargetEntry	*newTle;
+			ListCell	*lct;
+
+			Assert(IsA(lfirst(lcv), Var));
+
+			foreach(lct, plan->targetlist)
+			{
+				TargetEntry	*tle;
+
+				Assert(IsA(lfirst(lct), TargetEntry));
+
+				tle = (TargetEntry *) lfirst(lct);
+				if (IsA(tle->expr, Var) &&
+					(((Var *) (tle->expr))->varattno == ((Var *) lfirst(lcv))->varattno))
+					break;
+			}
+
+			if (!lct)
+			{
+				newTle = makeTargetEntry((Expr *) lfirst(lcv), targetResno,
+										 "" /* resname */, false /* resjunk */);
+				plan->targetlist = lappend(plan->targetlist, newTle);
+				++targetResno;
+			}
+		}
+		pcontext->found = true;
+
+		return;
+	}
+
+	if (plan->lefttree)
+	{
+		append_absent_targetlist_mutator(plan->lefttree, pcontext);
+
+		if (!pcontext->found && plan->righttree)
+			append_absent_targetlist_mutator(plan->righttree, pcontext);
+
+		if (pcontext->found)
+		{
+			ListCell	*lcv;
+			int			targetResno;
+
+			targetResno = list_length(plan->targetlist) + 1;
+
+			foreach(lcv, pcontext->absent_vars)
+			{
+				Var			*var;
+				ListCell	*lct;
+				TargetEntry *tle;
+
+				Assert(IsA(lfirst(lcv), Var));
+
+				var = (Var *) lfirst(lcv);
+
+				foreach(lct, plan->targetlist)
+				{
+					tle = (TargetEntry *) lfirst(lct);
+
+					if (IsA(tle->expr, Var) &&
+						((Var *) tle->expr)->varno == pcontext->result_relation_idx &&
+						((Var *) tle->expr)->varattno == var->varattno)
+						break;
+				}
+
+				if (!lct)
+				{
+
+					tle = makeTargetEntry(lfirst(lcv), targetResno++, "" /* resname */, false /* junk */);
+					plan->targetlist = lappend(plan->targetlist, tle);
+				}
+			}
+
+			return;
+		}
+	}
+
+	return;
+}
+
+
+/*
+ * Add absent vars to targetlist recursively. We just go into plan node, not subplan or
+ * other kind of nodes.
```

Comment:
I didn't quite understand what this function does. What vars are absent, from where?

Also, "append" makes me think of Append nodes, perhaps rename to "add_*" ?